### DISCUSSION TITLE
Error when trying to check Cell-Cycle impact in PCA

### DISCUSSION THREAD

**Discussion by Ima97** [satijalab/seurat#6894] [Q&A]
Nice to meet you.
I am new to bioinformatics.
Thank you for taking a look at this article.

I tried to do a Single cell RNA-Seq analysis using Seurat.

However, when I tried to check the effect of Cell-Cycle in PCA, I got the following error.

```
Error in irlba(A = t(x = object), nv = npcs, ...) : 
  max(nu, nv) must be positive
In addition: Warning message:
In PrepDR(object = object, features = features, verbose = verbose) :
  The following 97 features requested have not been scaled (running reduction without them): Mcm5, Pcna, Tyms, Fen1, Mcm7, Mcm4, Rrm1, Ung, Gins2, Mcm6, Cdca7, Dtl, Prim1, Uhrf1, Cenpu, Hells, Rfc2, Polr1b, Nasp, Rad51ap1, Gmnn, Wdr76, Slbp, Ccne2, Ubr7, Pold3, Msh2, Atad2, Rad51, Rrm2, Cdc45, Cdc6, Exo1, Tipin, Dscc1, Blm, Casp8ap2, Usp1, Clspn, Pola1, Chaf1b, Mrpl36, E2f8, Hmgb2, Cdk1, Nusap1, Ube2c, Birc5, Tpx2, Top2a, Ndc80, Cks2, Nuf2, Cks1b, Mki67, Tmpo, Cenpf, Tacc3, Pimreg, Smc4, Ccnb2, Ckap2l, Ckap2, Aurkb, Bub1, Kif11, Anp32e, Tubb4b, Gtse1, Kif20b, Hjurp, Cdca3, Jpt1, Cdc20, Ttk, Cdc25c, Kif2c, Rangap1, Ncapd2, Dlgap5, Cdca2, Cdca8, Ect2, Kif23, Hmmr, Aurka, Psrc1, Anln, Lbr, Ckap5, Cenpe, Ctcf, Nek2, G2e3, Gas2l3, Cbx5, Cenpa
```

I will also include the code for R at that time.

```
#Loading cell cycle marker list
s.genes <- sapply(cc.genes.updated.2019$s.genes, function(x){paste0(substr(x,1,1), tolower(substr(x,2,nchar(x))))})
g2m.genes <- sapply(cc.genes.updated.2019$g2m.genes, function(x){paste0(substr(x,1,1), tolower(substr(x,2,nchar(x))))})

#Checking the impact of Cell-Cycle on PCA
# Scaling the entire data
data.integrated <- ScaleData(object = data.integrated, features = rownames(data.integrated), verbose = FALSE)
# PC
data.integrated <- RunPCA(data.integrated, features = c(s.genes, g2m.genes))
```

We apologize for the inconvenience, but if anyone knows the cause of this error or how to resolve it, please let us know.

**Comment by gilstel** [comment]
I get the same error
Is there a solution?

**Comment by Gilgamesh-521** [comment]
I didn't get the error but a very similar warning (it can be regarded as an arror cuz nothing left for PCA after removing the warning genes). Has anyone found the reason?

```Warning in PrepDR5(object = object, features = features, layer = layer,  :
  The following features were not available: MCM5, FEN1, RRM1, UNG, GINS2, MCM6, CDCA7, DTL, PRIM1, UHRF1, HELLS, RFC2, RPA2, NASP, RAD51AP1, GMNN, WDR76, SLBP, CCNE2, UBR7, POLD3, MSH2, ATAD2, RAD51, EXO1, TIPIN, DSCC1, BLM, CASP8AP2, USP1, POLA1, CHAF1B, NDC80, CKS1B, TMPO, TACC3, AURKB, ANP32E, KIF20B, RANGAP1, NCAPD2, PSRC1, LBR, CKAP5, CTCF, G2E3, GAS2L3, CBX5, CENPA.
Warning in irlba(A = t(x = object), nv = npcs, ...) :
  You're computing too large a percentage of total singular values, use a standard svd instead.
Warning in irlba(A = t(x = object), nv = npcs, ...) :
  did not converge--results might be invalid!; try increasing work or maxit```

**Comment by rharao** [comment]
Replying on an old post in case this helps anyone who has the problem now:

> check the effect of Cell-Cycle in PCA

```r
data.integrated <- RunPCA(data.integrated, features = c(s.genes, g2m.genes))
```

In this function call, by setting `features = c(s.genes, g2m.genes)`, you're calculating the PCA using _only_ the cell cycle features as data for each cell; no other biological information is incorporated. Thus, the PCA will necessarily be _exclusively_ influenced by these cell cycle features.

I assume this is not what you want. Here's what you could do instead, starting after the scaling step:
* Calculate cell cycle scores for each cell: `data.integrated <- CellCycleScoring(data.integrated, s.genes, g2m.genes)`
* Calculate the PCA on the scaled features: `data.integrated <- RunPCA(data.integrated)
* Now, make some visualization that will allow you to identify if any PCs are strongly related to cell cycle state. There are a few ways you can do this.
  * `VizDimLoadings` or `DimHeatmap` can show if any of the top few PCs are strongly loaded with cell cycle features.
  * `PCAPlot`, or `VlnPlot` could also be applied for this; use the PCs as the coordinate variables and cell cycle phase assignment as grouping variables.
  * Get the relevant data with `Embeddings(data.integrated)` and `data.integrated[[]][c("S.Score", "G2M.Score")]` and plot them however you want.
